### PR TITLE
995x - Signposting fixes

### LINKED ISSUE #9952: Link headers for "license" are missing the <...> brackets
**What steps does it take to reproduce the issue?**

* When does this issue occur?

curl -v -D -L  https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/Z2JD58

* Which page(s) does it occurs on?

likely many

* What happens?

The header for the license is missing <...> brackets

link: <https://doi.org/10.7910/DVN/Z2JD58>;rel="cite-as", <https://dataverse.harvard.edu/api/access/datafile/7424560>;rel="item";type="application/octet-stream",<https://dataverse.harvard.edu/api/access/datafile/7424559>;rel="item";type="application/octet-stream",<https://dataverse.harvard.edu/api/access/datafile/7424558>;rel="item";type="application/octet-stream", <https://doi.org/10.7910/DVN/Z2JD58>;rel="describedby";type="application/vnd.citationstyles.csl+json",<https://dataverse.harvard.edu/api/datasets/export?exporter=schema.org&persistentId=doi:10.7910/DVN/Z2JD58>;rel="describedby";type="application/json+ld", <https://schema.org/AboutPage>;rel="type",<https://schema.org/Dataset>;rel="type", **_http://creativecommons.org/publicdomain/zero/1.0_;** rel="license", <https://dataverse.harvard.edu/api/datasets/:persistentId/versions/1.0/linkset?persistentId=doi:10.7910/DVN/Z2JD58> ; rel="linkset";type="application/linkset+json"

* To whom does it occur (all users, curators, superusers)?

* What did you expect to happen?

<http://creativecommons.org/publicdomain/zero/1.0>; rel="license"

**Which version of Dataverse are you using?**

current deployment at Harvard (5.14?)

**Any related open or closed issues to this bug report?**

not that I know

**Screenshots:**

No matter the issue, screenshots are always welcome.

To add a screenshot, please use one of the following formats and/or methods described here:

* https://help.github.com/en/articles/file-attachments-on-issues-and-pull-requests
*

### LINKED ISSUE #9953: Extraneous information in linkset JSON
**What steps does it take to reproduce the issue?**

* When does this issue occur?

curl -v -H "Accept: application/linkset+json" https://dataverse.harvard.edu/api/datasets/:persistentId/versions/1.0/linkset?persistentId=doi:10.7910/DVN/Z2JD58

* Which page(s) does it occurs on?

multiple

* What happens?

the "status" and "data" elements are not in the linkset JSON spec.

{"status":"OK","data":{"linkset":[{"anchor":"https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/Z2JD58","cite-as"......

* To whom does it occur (all users, curators, superusers)?

* What did you expect to happen?

{"linkset":[{"anchor":"https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/Z2JD58","cite-........

**Which version of Dataverse are you using?**

Deployment at Harvard (5.14?)

**Any related open or closed issues to this bug report?**

### LINKED ISSUE #9954: Dataset status change isn't visible after publish on auto refreshed dataset page (v5.14)
This behaviour only occurs in v5.14. It seems to be fixed in v6.0. 
However, we're looking for a quick fix which we can apply to v5.14.

There's a discussion on Zulip chat about this: https://dataverse.zulipchat.com/#narrow/stream/378866-troubleshooting/topic/dataset.20status.20change.20isn't.20visible.20after.20publish.20.28v5.2E14.29

Normally, when publishing a dataset, the dataset page refreshes every few seconds and then displays the published dataset.
With 5.14 however, the page is still refreshed, but at the end the old draft version is displayed, although the dataset is in status published. If you refresh the page manually, the message "_Info – The "DRAFT" version was not found. This is version "1.0"._ " is displayed. But if you go back to the list, and open the dataset again, it's in the correct state (published)

In the background the log shows a null pointer exception:
```
[2023-09-22T13:47:14.081+0200] [Payara 5.2022.5] [SCHWERWIEGEND] [] [javax.enterprise.resource.webcontainer.jsf.context] [tid: _ThreadID=98 _ThreadName=http-thread-pool::http-listener-1(2)] [timeMillis: 1695383234081] [levelValue: 1000] [[
    javax.faces.view.facelets.TagAttributeException: /dataset.xhtml @93,82 test="#{not empty DatasetPage.getSignpostingLinkHeader()}" /dataset.xhtml @93,82 test="#{not empty DatasetPage.getSignpostingLinkHeader()}": java.lang.NullPointerException
    at com.sun.faces.facelets.tag.TagAttributeImpl.getObject(TagAttributeImpl.java:318)
    ...
    Caused by: java.lang.NullPointerException
    at edu.harvard.iq.dataverse.DatasetPage.getSignpostingLinkHeader(DatasetPage.java:6149)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.base/java.lang.reflect.Method.invoke(Method.java:566)
    at javax.el.ELUtil.invokeMethod(ELUtil.java:236)
    ... 84 more
    ]]
```
The method DatasetPage.getSignPostingLinkHeader (https://github.com/IQSS/dataverse/blob/9f4ddbbbd3ea9aca23d2d828f575c7ba0747c2c7/src/main/java/edu/harvard/iq/dataverse/DatasetPage.java#L6148) is called a lot of times during the publishing process. The last 3 times it's called during the auto refresh phase, workingVersion is null. This leads to the null pointer exception. 

Fixing this method does not help, because then another NPE is thrown. The actual problem is that methods are called on DatasetPage when the page is refreshed, but workingVersion and dataset are null.

Here's a screencast showing this behaviour:
https://github.com/IQSS/dataverse/assets/5524896/c1262c31-5e12-4312-b0be-618a3112e102

Would be really helpful if anyone is aware which change from v5.14 to v6.0 fixedx this behaviour. 
Thank you!

### LINKED ISSUE #9957: Linkset contains links that are rejected  by the API
**What steps does it take to reproduce the issue?**

* When does this issue occur?

In a record's linkset, there are describedby links as follows;

```
<https://dataverse.harvard.edu/api/datasets/export?exporter=schema.org&persistentId=doi:10.7910/DVN/Z2JD58>
rel="describedby"
type="application/json+ld"
```

Attempting to retrieve this using curl 

```
curl -v -H "Accept: application/json+ld" "https://dataverse.harvard.edu/api/datasets/export?exporter=schema.org&persistentId=doi:10.7910/DVN/Z2JD58"
```
results in a 406 (API rejection) error

* Which page(s) does it occurs on?

* What happens?

406 error

* To whom does it occur (all users, curators, superusers)?

* What did you expect to happen?

Return me some yummy JSON :-)

**Which version of Dataverse are you using?**

current harvard dataverse release (5.14?)

**Any related open or closed issues to this bug report?**

**Screenshots:**

No matter the issue, screenshots are always welcome.

To add a screenshot, please use one of the following formats and/or methods described here:

* https://help.github.com/en/articles/file-attachments-on-issues-and-pull-requests
*

### PR DESCRIPTION
**Pull request by qqmyers**
**What this PR does / why we need it**: This addresses several issues raised by the community in signposting - two related to the format of output that wasn't per the spec, one regarding the mimetype used for a link, and one to avoid a potential issue where auto-updates of the page were failing in 5.14. (For the last one, it appears that 6.0 offers a fix. The fixes here will assure if workingversion == null for signposting occurs, it will not result in an exception that stops processing. They also add a check to make sure the versionId is set in the DatasetPage.refresh method, which was the problem in v5.14 as described in #9954.)

**Which issue(s) this PR closes**:

- Closes #9952
- Closes #9953 
- Closes #9954
- Closes #9957

**Special notes for your reviewer**:

**Suggestions on how to test this**: For the first two, read the issue and assure that the output now matches, i.e. <> chars exist in the license line and calling the linkset api doesn't return a linkset object as a child of a "data" json object. For the third, verify the mimetype used for the link in the issue is application/ld_json and verify the curl command in the issue - changed to use the new mimetype - works. For the last - as of now, regression testing that publishing a dataset works with the page finally updating to the published version is all we can do in 6.x. This is because 6.0 has changes that fix the known test cases and we don't know if the problem still exists in 6.x (with the fixes, if the problem does occur, we'll get waning messages in the log rather than having the dataset page display/refresh fail). FWIW, I did verify that the fixes here worked/works at QDR prior to the 6.0 fix.

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

**Is there a release notes update needed for this change?**:  added.

**Additional documentation**:

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/DatasetPage.java [hunk 1]
```diff
@@ -6144,7 +6144,7 @@ public String getWebloaderUrlForDataset(Dataset d) {
     String signpostingLinkHeader = null;
 
     public String getSignpostingLinkHeader() {
-        if (!workingVersion.isReleased()) {
+        if ((workingVersion==null) || (!workingVersion.isReleased())) {
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/DatasetPage.java:6147]: Is this a complete fix for #9954?

@haarli is reporting that with this null check in place another NPE emerges: https://github.com/IQSS/dataverse/issues/9954#issuecomment-1734924704

**Review by pdurbin — COMMENTED** [review]
For my reference, the original Signposting PR:

- #8981

I'm just asking a question in this review.

Related issues and PRs:

- #9941

File: src/main/java/edu/harvard/iq/dataverse/DatasetPage.java (hunk 1 above)
Comment (qqmyers) [reply src/main/java/edu/harvard/iq/dataverse/DatasetPage.java:6152]: It should be now (with the change in the refresh() method as well.) Note that the fixes don't appear to be needed in 6.x, but the null checks that are in this PR will avoid having the page fail if the problem recurs (and will throw a warning in the log).

Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/DatasetPage.java:6152]: Right. Fixed in c9c6cf2. Thanks.

**Review by pdurbin — COMMENTED** [review]
There's a failing test:

java.lang.NullPointerException: Cannot invoke "jakarta.json.JsonArray.getJsonObject(int)" because the return value of "jakarta.json.JsonObject.getJsonArray(String)" is null
	at edu.harvard.iq.dataverse.api.SignpostingIT.testSignposting(SignpostingIT.java:96)

See https://jenkins.dataverse.org/job/IQSS-Dataverse-Develop-PR/job/PR-9955/6/testReport/edu.harvard.iq.dataverse.api/SignpostingIT/testSignposting/

**Comment by pdurbin** [comment]
I added some docs and tests in 9f0b810. @qqmyers if I added any inaccuracies, please correct them!

I also merged the latest from develop.

**Review by pdurbin — APPROVED** [review]
Looks good!

**Pull request merged**